Mathlib/CategoryTheory/ChosenFiniteProducts.lean

```diff
@@ -0,0 +1,126 @@
+
+/-!
+
+# Categories with chosen finite products
+
+We introduce a class, `ChosenFiniteProducts`, which bundles explicit choices
+for a terminal object and binary products in a category `C`.
+This is primarily useful for categories which have finite products with good
+definitional properties, such as the category of types.
+
+Given a category with such an instance, we also provide the associated
+symmetric monoidal structure so that one can write `X ⊗ Y` for the explicit
+binary product and `𝟙_ C` for the explicit terminal object.
+
+-/
+
+namespace CategoryTheory
+
+universe v u
+
+/--
+An instance of `ChosenFiniteProducts C` bundles an explicit choice of a binary
+product of two objects of `C`, and a terminal object in `C`.
+
+Users should use the monoidal notation: `X ⊗ Y` for the product and `𝟙_ C` for
+the terminal object.
+-/
+class ChosenFiniteProducts (C : Type u) [Category.{v} C] where
+  /-- A choice of a limit binary fan for any two objects of the category.-/
+  product : (X Y : C) → Limits.LimitCone (Limits.pair X Y)
+  /-- A choice of a terminal object. -/
+  terminal : Limits.LimitCone (Functor.empty.{v} C)
+
+namespace ChosenFiniteProducts
+
+instance (C : Type u) [Category.{v} C] [ChosenFiniteProducts C] : MonoidalCategory C :=
+  monoidalOfChosenFiniteProducts ChosenFiniteProducts.terminal ChosenFiniteProducts.product
+
+instance (C : Type u) [Category.{v} C] [ChosenFiniteProducts C] : SymmetricCategory C :=
+  symmetricOfChosenFiniteProducts _ _
+
+variable {C : Type u} [Category.{v} C] [ChosenFiniteProducts C]
+
+open MonoidalCategory
+
+/--
+The unique map to the terminal object.
+-/
+def toUnit (X : C) : X ⟶ 𝟙_ C :=
+  terminal.isLimit.lift <| .mk _ <| .mk (fun x => x.as.elim) fun x => x.as.elim
+
+instance (X : C) : Unique (X ⟶ 𝟙_ C) where
+  default := toUnit _
+  uniq _ := terminal.isLimit.hom_ext fun ⟨j⟩ => j.elim
+
+/--
+This lemma follows from the preexisting `Unique` instance, but
+it is often convenient to use it directly as `apply toUnit_unique` forcing
+lean to do the necessary elaboration.
+-/
+lemma toUnit_unique {X : C} (f g : X ⟶ 𝟙_ _) : f = g :=
+  Subsingleton.elim _ _
+
+/--
+Construct a morphism to the product given its two components.
+-/
+def lift {T X Y : C} (f : T ⟶ X) (g : T ⟶ Y) : T ⟶ X ⊗ Y :=
+  (product X Y).isLimit.lift <| Limits.BinaryFan.mk f g
+
+/--
+The first projection from the product.
+-/
+def fst (X Y : C) : X ⊗ Y ⟶ X :=
+  letI F : Limits.BinaryFan X Y := (product X Y).cone
+  F.fst
+
+/--
+The second projection from the product.
+-/
+def snd (X Y : C) : X ⊗ Y ⟶ Y :=
+  letI F : Limits.BinaryFan X Y := (product X Y).cone
+  F.snd
+
+@[reassoc (attr := simp)]
+lemma lift_fst {T X Y : C} (f : T ⟶ X) (g : T ⟶ Y) : lift f g ≫ fst _ _ = f := by
+  simp [lift, fst]
+
+@[reassoc (attr := simp)]
+lemma lift_snd {T X Y : C} (f : T ⟶ X) (g : T ⟶ Y) : lift f g ≫ snd _ _ = g := by
+  simp [lift, snd]
+
+@[ext 1050]
+lemma lift_unique {T X Y : C} (f g : T ⟶ X ⊗ Y)
```

Comment:
I would prefer the name `hom_ext`, simply because `lift` doesn't actually appear in the statement.